circuit-construction/src/writer.rs

```diff
@@ -744,6 +770,17 @@ impl<F: PrimeField> Cs<F> for WitnessGenerator<F> {
             None
         }
     }
+
+    fn cached_constants(&mut self, x: F) -> Var<F> {
```

Comment:
This method seems to behave exactly the same both for WitnessGenerator and for System. Perhaps you can move it to the trait itself to avoid duplication.

Reply:
The problem is that I can't access `self.cached_constants` because the trait doesn't know any of the fields of `Self`.

That's why I think we should move to a single struct instead of two, this is really redundant code